File: rust/src/transport.rs
use std::time::Duration;

use reqwest::{Client, Method, Response};
use serde_json::Value;

use crate::config::ConnectionConfig;
use crate::error::{Error, Result};

#[derive(Clone)]
pub(crate) struct ControlClient {
    config: ConnectionConfig,
    client: Client,
}

impl ControlClient {
    pub(crate) fn new(config: ConnectionConfig) -> Result<Self> {
        let client = Client::builder()
            .timeout(Duration::from_secs(config.request_timeout_secs))
            .build()?;
        Ok(Self { config, client })
    }

    pub(crate) async fn get(&self, path: &str) -> Result<Value> {
        self.request(Method::GET, path, None).await
    }

    pub(crate) async fn post(&self, path: &str, body: Value) -> Result<Value> {
        self.request(Method::POST, path, Some(body)).await
    }

    pub(crate) async fn post_idempotent(
        &self,
        path: &str,
        body: Value,
        idempotency_key: Option<&str>,
    ) -> Result<Value> {
        self.request_with_idempotency(Method::POST, path, Some(body), idempotency_key)
            .await
    }

    pub(crate) async fn put(&self, path: &str, body: Value) -> Result<Value> {
        self.request(Method::PUT, path, Some(body)).await
    }

    pub(crate) async fn patch(&self, path: &str, body: Value) -> Result<Value> {
        self.request(Method::PATCH, path, Some(body)).await
    }

    pub(crate) async fn delete(&self, path: &str) -> Result<Value> {
        self.request(Method::DELETE, path, None).await
    }

    pub(crate) async fn delete_with_body(&self, path: &str, body: Value) -> Result<Value> {
        self.request(Method::DELETE, path, Some(body)).await
    }

    async fn request(&self, method: Method, path: &str, body: Option<Value>) -> Result<Value> {
        self.request_with_idempotency(method, path, body, None)
            .await
    }

    async fn request_with_idempotency(
        &self,
        method: Method,
        path: &str,
        body: Option<Value>,
        idempotency_key: Option<&str>,
    ) -> Result<Value> {
        let api_key = self.config.api_key.as_ref().ok_or(Error::MissingApiKey)?;
        let mut request = self
            .client
            .request(method, join_url(&self.config.api_url, path))
            .bearer_auth(api_key);
        if let Some(idempotency_key) = idempotency_key {
            request = request.header("Idempotency-Key", idempotency_key);
        }
        if let Some(body) = body {
            request = request.json(&body);
        }
        json_response(request.send().await?).await
    }
}

#[derive(Clone)]
pub(crate) struct DataPlaneClient {
    pub(crate) base_url: String,
    pub(crate) token: String,
    client: Client,
}

impl DataPlaneClient {
    pub(crate) fn new(base_url: String, token: String, config: &ConnectionConfig) -> Result<Self> {
        let client = Client::builder()
            .timeout(Duration::from_secs(config.request_timeout_secs))
            .build()?;
        Ok(Self {
            base_url,
            token,
            client,
        })
    }

    pub(crate) async fn get_json(&self, path: &str) -> Result<Value> {
        json_response(
            self.client
                .get(join_url(&self.base_url, path))
                .bearer_auth(&self.token)
                .send()
                .await?,
        )
        .await
    }

    pub(crate) async fn post_json(&self, path: &str, body: Value) -> Result<Value> {
        json_response(
            self.client
                .post(join_url(&self.base_url, path))
                .bearer_auth(&self.token)
                .json(&body)
                .send()
                .await?,
        )
        .await
    }

    pub(crate) async fn delete_json(&self, path: &str) -> Result<Value> {
        json_response(
            self.client
                .delete(join_url(&self.base_url, path))
                .bearer_auth(&self.token)
                .send()
                .await?,
        )
        .await
    }

    pub(crate) async fn get_bytes_with_limit(
        &self,
        path: &str,
        max_bytes: Option<usize>,
    ) -> Result<Vec<u8>> {
        let response = self
            .client
            .get(join_url(&self.base_url, path))
            .bearer_auth(&self.token)
            .send()
            .await?;
        if !response.status().is_success() {
            return Err(Error::from_status(
                response.status(),
                &read_json_or_text(response).await?,
            ));
        }
        read_limited_bytes(response, max_bytes).await
    }

    pub(crate) async fn put_bytes(&self, path: &str, data: Vec<u8>) -> Result<Value> {
        json_response(
            self.client
                .put(join_url(&self.base_url, path))
                .bearer_auth(&self.token)
                .header("content-type", "application/octet-stream")
                .body(data)
                .send()
                .await?,
        )
        .await
    }
}

async fn read_limited_bytes(mut response: Response, max_bytes: Option<usize>) -> Result<Vec<u8>> {
    if let (Some(content_length), Some(max_bytes)) = (response.content_length(), max_bytes) {
        if content_length > max_bytes as u64 {
            return Err(Error::ByteLimitExceeded {
                stream: "file",
                max_bytes,
                actual_bytes: content_length as usize,
            });
        }
    }

    let mut bytes = Vec::new();
    while let Some(chunk) = response.chunk().await? {
        if let Some(max_bytes) = max_bytes {
            let actual_bytes = bytes.len().saturating_add(chunk.len());
            if actual_bytes > max_bytes {
                return Err(Error::ByteLimitExceeded {
                    stream: "file",
                    max_bytes,
                    actual_bytes,
                });
            }
        }
        bytes.extend_from_slice(&chunk);
    }
    Ok(bytes)
}

async fn json_response(response: Response) -> Result<Value> {
    let status = response.status();
    let payload = read_json_or_text(response).await?;
    if !status.is_success() {
        return Err(Error::from_status(status, &payload));
    }
    Ok(payload)
}

async fn read_json_or_text(response: Response) -> Result<Value> {
    let text = response.text().await?;
    if text.trim().is_empty() {
        return Ok(Value::Object(Default::default()));
    }
    Ok(serde_json::from_str(&text).unwrap_or_else(|_| serde_json::json!({ "message": text })))
}

pub(crate) fn join_url(base: &str, path: &str) -> String {
    format!(
        "{}{}",
        base.trim_end_matches('/'),
        if path.starts_with('/') {
            path.to_string()
        } else {
            format!("/{path}")
        }
    )
}

File: ts/src/transport.ts
import { ConnectionConfig } from './connectionConfig.js'
import { AuthenticationError, TimeoutError, errorFromResponse } from './errors.js'

type JsonRecord = Record<string, unknown>

export class ControlClient {
  constructor(private readonly config: ConnectionConfig) {}

  get(path: string, opts: RequestOpts = {}) {
    return this.request(path, { ...opts, method: 'GET' })
  }

  post(path: string, opts: RequestOpts = {}) {
    return this.request(path, { ...opts, method: 'POST' })
  }

  put(path: string, opts: RequestOpts = {}) {
    return this.request(path, { ...opts, method: 'PUT' })
  }

  patch(path: string, opts: RequestOpts = {}) {
    return this.request(path, { ...opts, method: 'PATCH' })
  }

  delete(path: string, opts: RequestOpts = {}) {
    return this.request(path, { ...opts, method: 'DELETE' })
  }

  private async request(path: string, opts: RequestOpts): Promise<JsonRecord> {
    if (!this.config.apiKey) {
      throw new AuthenticationError('WATASU_API_KEY is required')
    }

    const response = await fetchWithTimeout(
      joinUrl(this.config.apiUrl, path),
      {
        method: opts.method,
        headers: {
          ...this.config.authHeaders,
          ...(opts.json ? { 'content-type': 'application/json' } : {}),
        },
        body: opts.json ? JSON.stringify(opts.json) : undefined,
      },
      opts.requestTimeoutMs ?? this.config.requestTimeoutMs
    )

    return parseJsonResponse(response)
  }
}

export class DataPlaneClient {
  constructor(
    readonly baseUrl: string,
    readonly token: string,
    private readonly config: ConnectionConfig
  ) {}

  getJson(path: string, opts: RequestOpts = {}) {
    return this.request(path, { ...opts, method: 'GET' }) as Promise<JsonRecord>
  }

  postJson(path: string, opts: RequestOpts = {}) {
    return this.request(path, { ...opts, method: 'POST' }) as Promise<JsonRecord>
  }

  deleteJson(path: string, opts: RequestOpts = {}) {
    return this.request(path, { ...opts, method: 'DELETE' }) as Promise<JsonRecord>
  }

  async getBytes(path: string, opts: RequestOpts = {}): Promise<Uint8Array> {
    const response = await this.raw(path, { ...opts, method: 'GET' })
    return new Uint8Array(await response.arrayBuffer())
  }

  async putJson(path: string, data: BodyInit | Uint8Array, opts: RequestOpts = {}) {
    return parseJsonResponse(
      await this.raw(path, {
        ...opts,
        method: 'PUT',
        body: data,
        headers: { 'content-type': 'application/octet-stream', ...(opts.headers ?? {}) },
      })
    )
  }

  async raw(path: string, opts: RequestOpts): Promise<Response> {
    const response = await fetchWithTimeout(
      joinUrl(this.baseUrl, path),
      {
        method: opts.method,
        headers: {
          Authorization: `Bearer ${this.token}`,
          ...(opts.json ? { 'content-type': 'application/json' } : {}),
          ...(opts.headers ?? {}),
        },
        body: opts.json ? JSON.stringify(opts.json) : (opts.body as BodyInit | undefined),
      },
      opts.requestTimeoutMs ?? this.config.requestTimeoutMs
    )

    if (!response.ok) {
      throw errorFromResponse(response.status, await readJsonOrText(response))
    }
    return response
  }

  private async request(path: string, opts: RequestOpts): Promise<unknown> {
    return parseJsonResponse(await this.raw(path, opts))
  }
}

export interface RequestOpts {
  method?: string
  json?: unknown
  body?: BodyInit | Uint8Array
  headers?: Record<string, string>
  requestTimeoutMs?: number
}

export function withQuery(path: string, params: Record<string, string | number | boolean | undefined>) {
  const url = new URL(path, 'https://placeholder.local')
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined) url.searchParams.set(key, String(value))
  }
  return `${url.pathname}${url.search}`
}

async function parseJsonResponse(response: Response): Promise<JsonRecord> {
  const payload = await readJsonOrText(response)
  if (!response.ok) {
    throw errorFromResponse(response.status, payload)
  }
  return payload && typeof payload === 'object' ? (payload as JsonRecord) : {}
}

async function readJsonOrText(response: Response): Promise<unknown> {
  const text = await response.text()
  if (!text) return {}
  try {
    return JSON.parse(text) as unknown
  } catch {
    return { message: text }
  }
}

function fetchWithTimeout(url: string, init: RequestInit, timeoutMs: number): Promise<Response> {
  const controller = new AbortController()
  const timeout = setTimeout(() => controller.abort(), timeoutMs)
  return fetch(url, { ...init, signal: controller.signal }).catch((error) => {
    if (error?.name === 'AbortError') throw new TimeoutError()
    throw error
  }).finally(() => clearTimeout(timeout))
}

function joinUrl(base: string, path: string): string {
  const normalizedBase = base.replace(/\/+$/, '')
  const normalizedPath = path.startsWith('/') ? path : `/${path}`
  return `${normalizedBase}${normalizedPath}`
}
